### PR TITLE
Rust agent developer guide documentation 

### PR DESCRIPTION
**Pull request by SirCipher**

### CODE REVIEW

File: src/_rust_devguide/2.1-project-setup.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  - /rust/developer-guide/project-setup.html
+---
+
+To initiate the setup of a SwimOS project, start by creating and navigating to the directory where it will be located. First, lets start by creating a new server project named `server`:
+
+```shell
+cargo new server --bin
+```
+
+Then we need to declare our dependencies. Let's start by adding the following to our `Cargo.toml`:
```
Comment (horned-sphere) [inline src/_rust_devguide/2.1-project-setup.md:18]: I don't really like the conversational tone. I think would be better phrased in the imperative ("do this" rathern than "we should do this"). Just personal taste though and I don't know what the style in the rest of the docs is.

File: src/_rust_devguide/2.1-project-setup.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+```
+
+The SwimOS platform is built using the `tokio` runtime so we need to include it.
+
+Lets add the required imports for our example application to the `main.rs` file:
+
+```rust
+use swim::{
+    agent::{
+        agent_lifecycle::utility::HandlerContext,
```
Comment (horned-sphere) [inline src/_rust_devguide/2.1-project-setup.md:34]: Listing the imports up front without context feels odd. I know it would be a pain but I think it would flow better if they are introduced as you add the actual code that needs them.

File: src/_rust_devguide/2.2-hello-agent.md [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+redirect_from:
+  - /rust/developer-guide/hello-agent
+  - /rust/developer-guide/hello-agent.html
+---
+
+Having completed the setup, we're ready to continue our tutorial application. We'll now incorporate SwimOS's fundamental building blocks - agents.
+
+# Agents
+
+A Stateful Object is a simple structural representation of an entity that lives amongst your streaming data. A Stateful Object continuously consumes updates of synced state and for each update, it computes new state that automatically gets synced to listeners. In SwimOS, these Stateful Objects are named Agents and the state contained within them is exposed through Lanes; these are similar to `Durable State` in Akka and `State` in Flink but the state is addressable using a URI relative to an instance of an Agent.
```
Comment (horned-sphere) [inline src/_rust_devguide/2.2-hello-agent.md:16]: I'm not sure I'd use the word object as it has associations with OOP which there isn't any of here.

Also, I wouldn't assume familiarity with Flink and Akka (they're both fairly niche). It would be reasonable to have a section where you compare Swim to other frameworks but that should be a section on its own rather than mixed in with the tutorial.

File: src/_rust_devguide/2.2-hello-agent.md [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+Agents are defined using structs that can be automatically derived using the declared macros below. Lets define one in our `server/src/main.rs` file:
+
+```rust
+#[derive(AgentLaneModel)]
+#[projections]
```
Comment (horned-sphere) [inline src/_rust_devguide/2.2-hello-agent.md:22]: I think you should address what any proc macros do (in summary rather than in detail) as they are introduced. I get annoyed with "magical" elements are dropped into the code without explanation.

File: src/_rust_backend/agent-design.md [hunk 5]
```diff
@@ -14,11 +14,11 @@ In this guide we will discuss some best practices and ideas for designing and mo
 
 ## Entity Modeling
 
-**Entities** are identifiable domain elements (the nouns) and Swim **nodes** are addressable groupings of stateful agents which perform some logic.
+**Entities** are identifiable domain elements and Swim **nodes** are addressable groupings of stateful agents which perform some logic.
 Comparing the two definitions, we get a pretty good idea of how to map our entities to nodes, (almost) one-to-one.
 
 Anything **uniquely addressable** can be mapped to a node.
-This can be singleton such as a node representing the whole system (`/system`), some entity type with a globally unique id (`/vehicle/:id`), or entities where some composite unique address can be constructed (`/building/:buildingId/room/:roomId`, since roomId is only unique to each building).
+This can be singleton such as a node representing the whole system (`/system`), some entity type with a globally unique id (`/vehicle/:id`), or entities where some composite unique address can be constructed (`/building/:buildingId/room/:roomId`, since `roomId` is only unique to each building).
```
Comment (horned-sphere) [inline src/_rust_backend/agent-design.md:21]: You should clarify that the second kind defines a (potentially) infinite family.

File: src/_rust_devguide/2.2-hello-agent.md [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+
+/// Here we bind this struct's implementation to our agent above.
+#[lifecycle(ExampleAgent)]
+impl ExampleLifecycle {
+    /// This method will be invoked when the agent starts.
+    #[on_start]
+    pub fn on_start(
+        &self,
+        context: HandlerContext<ExampleAgent>,
+    ) -> impl EventHandler<ExampleAgent> {
```
Comment (horned-sphere) [inline src/_rust_devguide/2.2-hello-agent.md:40]: You should describe how event handlers work before going into a full lifecycle derivation.

File: src/_rust_devguide/2.1-project-setup.md (hunk 2 above)
Comment (SirCipher) [reply src/_rust_devguide/2.1-project-setup.md:34]: Are you suggesting that for each code block that introduces a new import it's added in the block too? Something like this:

```rust
use swim_form::Form;

#[derive(Debug, Form, Copy, Clone)]
pub enum Operation {
    Add(i64),
    Sub(i64),
}

File: src/_rust_devguide/2.2-hello-agent.md (hunk 3 above)
Comment (SirCipher) [reply src/_rust_devguide/2.2-hello-agent.md:16]: I'll reword it to remove the word object. R.e, Flink/Akka, I could prefix it with "if you're familiar with..."

File: src/_rust_devguide/2.2-hello-agent.md (hunk 4 above)
Comment (SirCipher) [reply src/_rust_devguide/2.2-hello-agent.md:22]: I did wonder this too (and feel similar) but it's a bit of a Pandora's box as this page was just an introduction to getting something working and showing how things work. The subsequent sections will start to expand on these things: https://github.com/swimos/swimos-docs/pull/173/files#diff-11006668302b80fc07fd0c5c9c7d076cddcab4168fdc8ec4c571534437a5ac0fR67

File: src/_rust_devguide/2.1-project-setup.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        event_handler::{EventHandler, HandlerActionExt},
+        lanes::ValueLane,
+        lifecycle, projections, AgentLaneModel,
+    },
+    route::RoutePattern,
+    server::{Server, ServerBuilder, ServerHandle},
+};
+```
+
+Now lets add our client project named `client`:
```
Comment (PiggottDev) [inline src/_rust_devguide/2.1-project-setup.md:45]: Creating a client seems a little fast, I think just getting a working server should be the first step, interacting via a client can come in a following section.

File: src/_rust_devguide/2.2-hello-agent.md [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+        let n = *value;
+        context.effect(move || {
+            println!("Setting value to: {}", n);
+        })
+    }
+}
+
+
+```
```
Comment (PiggottDev) [inline src/_rust_devguide/2.2-hello-agent.md:77]: Personally I think this introduces too many Swim concepts in the first code snippet (agents, lanes, lifecycles, URIs), I would like it to be more iterative so you can see the impact of each line of code.

**Review by PiggottDev — COMMENTED** [review]
I think it's good but too fast.

I would like to have a runnable server as soon as possible - even if it does nothing but log something -  so I can run the code and see how each line is affecting my output.

File: src/_rust_devguide/2.2-hello-agent.md [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+# Starting the server
+
+To build and run a server, we first need to map the schema of agents to a `ServerBuilder` and configure any properties.
+
+To register agents with a `ServerBuilder`, we need to associate the URIs of the agents with the agents. We do this as follows in `server/src/main.rs`:
+
+```rust
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn Error + Send + Sync>> {
+    // Here we create a dynamic route named "/example/:name" to use for our `ExampleAgent`.
```
Comment (DobromirM) [inline src/_rust_devguide/2.2-hello-agent.md:96]: Maybe a brief explanation about dynamic routes will be good here. Especially how they are formed, how many dynamic components they can have, where they can appear, etc.
